cli/flox/src/commands/containerize/macos_containerize_proxy.rs

```diff
@@ -34,26 +47,79 @@ impl ContainerizeProxy {
-        _name: impl AsRef<str>,
-        tag: impl AsRef<str>,
-    ) -> Result<ContainerSource, Self::Error> {
-        // Inception L1: Container runtime args.
+    /// Base command for the container runtime.
+    fn runtime_base_command(&self) -> Command {
         let mut command = self.container_runtime.to_command();
         command.arg("run");
         command.arg("--rm");
+        command
+    }
+
+    // Use a Nix container that matches the version of Nix that this Flox
+    // has been built with because it's smaller and changes less frequently
+    // than a Flox container of the corresponding version, which result in less
+    // container image pulls. It also prevents the chicken-and-egg problem when
+    // we bump `VERSION` in Flox but haven't published the container image yet.
+    fn container_image(&self) -> String {
+        format!(
+            "{}:{}",
+            NIX_PROXY_IMAGE,
+            FLOX_CONTAINERIZE_PROXY_IMAGE_REF
+                .clone()
+                .unwrap_or(NIX_VERSION.clone())
+        )
+    }
+
+    /// Add a cache volume mount to the container runtime command.
+    fn add_cache_mount(&self, command: &mut Command, path: &str) {
+        command.args([
+            "--mount",
+            &format!("type=volume,src={},dst=/{}", CONTAINER_VOLUME, path),
+        ]);
+    }
+
+    /// Copy the Nix store from the container image to the cache volume.
+    fn populate_cache_volume(&self) -> Result<(), ContainerizeProxyError> {
+        let mut command = self.runtime_base_command();
+
+        // The cache volume has to be mounted in parallel to the container's own
+        // `/nix` and at a prefix where it can be treated as a new local root:
+        // https://nix.dev/manual/nix/2.24/command-ref/new-cli/nix3-help-stores#local-store
+        let cache_root = "/cache";
+        self.add_cache_mount(&mut command, &format!("{cache_root}/nix"));
+
+        command.arg(self.container_image());
+        // Profiles have to be copied too because the mount will shadow the
+        // container's `/nix/var/nix/profiles` and break `PATH`.
+        command.args(["bash", "-c", &formatdoc! {"
+            set -euo pipefail
+            nix --extra-experimental-features nix-command copy --all --no-check-sigs --to {cache_root}
+            cp -R /nix/var/nix/profiles {cache_root}/nix/var/nix/
+        "}]);
+
+        command.stdout(Stdio::piped());
+        command.stderr(Stdio::piped());
+        debug!(?command, "Populating container cache volume");
+
+        let output = command
+            .output()
+            .map_err(ContainerizeProxyError::PopulateCacheVolume)?;
+
+        if !output.status.success() {
+            return Err(ContainerizeProxyError::PopulateCacheVolume(
+                std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    String::from_utf8_lossy(&output.stderr).to_string(),
+                ),
+            ));
+        }
+
```

Comment:
suggestion: tap into output?

Reply:
Using `utils::WireTap`? Can you remind me why it would be useful in this case?

Reply:
Yeah, it wouldn't hurt for debugging. I've pushed a fixup commit. Can you double check that I'm holding it correctly?

Reply:
Looks correct 👍🏼